### PR TITLE
If necessary, prompt user to restart after add-on installation when using mozAddonManager (closes #1360).

### LINKED ISSUE #1360: Implement in product restart Prompt for e10s users in 49
Per: https://bugzilla.mozilla.org/show_bug.cgi?id=1302846

The issue is that we need to get all Test Pilot experiments on the e10s whitelist for FF49 if we want e10s users to avoid browser restart. If we are unable to make this list, we need a graceful fallback. Fortunately, mozAddonManager allows for [in-content restart UI](https://github.com/mozilla/addons-frontend/pull/568).

@chuckharmston here's a provisional mock up. I'm not 100% wedded to the design, but the idea here is to rive users inexorably toward that restart button.

@andymckay has offered to provide support in implementation here. 

e10s restart

### PR DESCRIPTION
**Pull request by chuckharmston**
~~This is WIP, getting it out there so it's easy to follow progress.~~
## Screencaps

As of ab498de, here's what the e10s experience looks like:
- [In 49 with e10s, where a restart is required](https://vid.me/rrPw)
- [In 50 with e10s, where a restart is not required](https://vid.me/LLC8)

As of d1cced6, showing the two reboot flows:
- [When installing from the landing page](https://vid.me/4Mq9)
- [When installing from an experiment page](https://vid.me/Wm3N)
## TODO
- [x] Detect necessary post-install restarts.
- [x] Prompt user for restart if necessary.
- [x] Make the prompt look [like this](https://cloud.githubusercontent.com/assets/3323249/18564740/3965abe2-7b8d-11e6-95fe-1e62d07c0930.png).
- ~~On add-on first run, if there are no open test Test Pilot tabs, open one.~~ (#1374)
- ~~Instrument this process (TBD @johngruen @ckprice)~~ (#1375)
## Testing this

To test this flow locally, you'll need a few things:
- Firefox === 49
- New profile
- No add-ons installed
- `browser.tabs.remote.force-enable` set to `false`
- `browser.tabs.remote.autostart` set to `true`
- You must use a signed copy of the add-on. Download the XPI from the production site, and move it to `/frontend/build/static/addon/addon.xpi`. You may need to do this multiple times as the gulp task rebuilds the site.
- The [full SSL proxy setup from the `mozAddonManager` PR](https://github.com/mozilla/testpilot/pull/1330#issuecomment-245821762).

You can verify that e10s is enabled by going to about:support and looking for the `Multiprocess Windows` value; you want it to be greater than 0.

### CODE REVIEW

File: frontend/src/app/components/App.js
Comment (chuckharmston) [inline frontend/src/app/components/App.js:18]: Ideally this is the only connected component, and the entirety of the redux store is proliferated to children via props.

It's orthogonal to this PR so I didn't implement it here, but we would likely do it like this:

```
{React.cloneElement(this.props.children, ...this.props)}
```

File: frontend/src/app/components/MainInstallButton.js
Comment (chuckharmston) [inline frontend/src/app/components/MainInstallButton.js:18]: This is the sort of thing I'd like to get rid of in #1367; there's no reason this value shouldn't be on the Redux store and available for other components to access.

File: frontend/src/app/components/MainInstallButton.js
Comment (chuckharmston) [inline frontend/src/app/components/MainInstallButton.js:63]: This is how you get access to the store object. [More info](https://egghead.io/lessons/javascript-redux-passing-the-store-down-implicitly-via-context)

File: frontend/src/app/index.js
Comment (chuckharmston) [inline frontend/src/app/index.js:51]: https://addons.mozilla.org/en-US/firefox/addon/remotedev/ 💓

File: frontend/src/app/lib/router.js
Comment (chuckharmston) [inline frontend/src/app/lib/router.js:21]: This is how you might use `<App>`; When matching, its `children` prop is set to any nested components that also match. This works infinitely deep, and allows for some really cool DRY patterns.

File: frontend/src/app/actions/addon.js
Comment (chuckharmston) [inline frontend/src/app/actions/addon.js:6]: Filed https://github.com/mozilla/testpilot/issues/1371 to use constants for these.

**Comment by johngruen** [comment]
@chuckharmston dont' sweat the look of the 'reload' page UI too much.

**Comment by johngruen** [comment]
@ckprice please see initial comment regarding instrumentation

**Comment by chuckharmston** [comment]
> @chuckharmston dont' sweat the look of the 'reload' page UI too much.

How's this, @johngruen? (+ different copy when installed from an add-on page)

screen shot 2016-09-18 at 1 23 43 pm

screen shot 2016-09-18 at 1 23 55 pm

I don't think we'd want to do the button-looking non-button.

**Comment by chuckharmston** [comment]
Updated initial comment with videos from d1cced6.

**Comment by chuckharmston** [comment]
For the sake of smaller PRs and quicker testing, I moved two orthogonal pieces of this into their own bugs (#1374, #1375) and am ready for this to be reviewed.

**Comment by chuckharmston** [comment]
Added some testing notes to the PR description; it's a doozy.

**Comment by lmorchard** [comment]
Taking a look

File: frontend/tasks/server.js
Comment (lmorchard) [inline frontend/tasks/server.js:5]: Nitpick: Didn't want to use this module here, because we're generating static `index.html` landing pages for S3. This can mask missing landing pages that we forgot to add to the generation task. Maybe more suitable for a testing scenario to catch?

File: frontend/src/app/components/MainInstallButton.js
Comment (lmorchard) [reply frontend/src/app/components/MainInstallButton.js:63]: I watched that video, but I thought the context system was undocumented and not to be relied on? [The docs have a big scary caveat on it](https://facebook.github.io/react/docs/context.html), though I know that's how Redux passes things around with the `<Provider>` / `connect` abstraction. Solves the issue I was having with passing things through the router, though

File: frontend/src/app/index.js
Comment (lmorchard) [reply frontend/src/app/index.js:51]: :+1: 

Nitpick: worth adding a note to dev docs about what this enables, but I wouldn't block this PR for it.

File: frontend/src/app/actions/addon.js
Comment (lmorchard) [reply frontend/src/app/actions/addon.js:6]: Commented over there

File: frontend/src/app/components/MainInstallButton.js
Comment (lmorchard) [reply frontend/src/app/components/MainInstallButton.js:63]: This also couples the component to Redux, right? Seems to make it harder to keep this presentation-only and [use `connect` to pass in the mapped store data & dispatch helpers in props](https://github.com/mozilla/testpilot/blob/b33a19b08295b7d6ef0e789afb6ea1ccfb31e315/frontend/src/app/containers/ExperimentPage.js#L10). That seemed like an important pattern to keep things loosely bound & easily testable. Wondering if we can somehow get this stuff fed in from the `App` component as props?

File: frontend/src/app/containers/Restart.js
Comment (lmorchard) [inline frontend/src/app/containers/Restart.js:1]: Nitpick: Should this instead live with the rest of the presentation-only components under `components`?

File: frontend/src/app/lib/addon.js
Comment (lmorchard) [inline frontend/src/app/lib/addon.js:20]: Does this need to be paired with a `removeEventListener` once we've finished processing here? Guessing the two subscriptions below get garbage collected after the install in question is over, but would this listener catch _all_ install progress events going forward? Probably harmless, but concerned about "dead" code getting called.

File: frontend/src/app/lib/addon.js
Comment (lmorchard) [inline frontend/src/app/lib/addon.js:89]: (derp.)

File: frontend/src/app/lib/addon.js
Comment (lmorchard) [inline frontend/src/app/lib/addon.js:109]: (herp derp.)

File: frontend/src/app/components/MainInstallButton.js
Comment (lmorchard) [reply frontend/src/app/components/MainInstallButton.js:63]: Actually, is this what you meant with this bit in `<App>`?

``` javascript
{React.cloneElement(this.props.children, ...this.props)}
```

Could we supply a generic `installAddon` helper function in a prop from `<App>` and leave this component ignorant of the Redux store? Happy not to block this PR and save that as a future improvement if so.

File: frontend/src/app/components/App.js
Comment (lmorchard) [reply frontend/src/app/components/App.js:18]: I think I grok that, and that sounds like a good follow-on to this!

**Review by lmorchard — COMMENTED** [review]
Works for me, code looks good. My only things are nitpicks and architectural questions not worth blocking the PR to address.

**Review by lmorchard — APPROVED** [review]
Looks like I picked "comment" and not "approve", but I can't change it. Anyway, [this was my real review summary](https://github.com/mozilla/testpilot/pull/1369#pullrequestreview-576724).

File: frontend/src/app/components/MainInstallButton.js
Comment (chuckharmston) [reply frontend/src/app/components/MainInstallButton.js:63]: I tried to keep the surface area of this PR small because of its urgency, but yep, that's exactly what the promised land looks like.

File: frontend/src/app/containers/Restart.js
Comment (chuckharmston) [reply frontend/src/app/containers/Restart.js:1]: For the sake of having an easy rule, I'd suggest putting anything that gets handed to the router in `containers`, and anything else in `components`.

File: frontend/tasks/server.js
Comment (chuckharmston) [reply frontend/tasks/server.js:5]: What's the generation process? Are there docs for that? I did this because the route I added for `/restart` wasn't accessible `¯\_(ツ)_/¯`

Comment (lmorchard) [reply frontend/tasks/server.js:5]: Ah, yeah docs would be nice, but there are none. So, this is actually something that needs fixing or else the static deploy will result in a 404.

New paths need to be added to [this list over here in the pages task](https://github.com/mozilla/testpilot/blob/master/frontend/tasks/pages.js#L25). (which I missed, and _should_ have been my review comment :) )

File: frontend/src/app/components/MainInstallButton.js
Comment (lmorchard) [reply frontend/src/app/components/MainInstallButton.js:63]: Okay, cool beans. Was worried I missed something, but nope. Let's save this for future work

**Comment by lmorchard** [comment]
Un-assigning myself, because I think all my concerns are addressed. Good to merge after a rebase IMO

**Pull request merged**
